File: java/depth-first-search/CountConnectedComponents.java
import java.util.ArrayList;
import java.util.List;

class Solution {

    public List<List<Integer>> buildGraph(int[][] edges, int n) {
        List<List<Integer>> graph = new ArrayList<>();
        
        for(int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for(int[] edge : edges) {
            graph.get(edge[0]).add(edge[1]);
            graph.get(edge[1]).add(edge[0]);
        }

        return graph;

    }

    private void solve(int node, List<List<Integer>> graph, boolean[] visited) {
        visited[node] = true;
        for(int nbr : graph.get(node)) {
            if(!visited[nbr]) {
                solve(nbr, graph, visited);
            }
        }
    }

    public int countComponents(int n, int[][] edges) {
        List<List<Integer>> graph = buildGraph(edges, n);
        boolean[] visited = new boolean[n];
        int count = 0;
        for(int src = 0; src < n; src++) {
            if(!visited[src]) {
                solve(src, graph, visited);
                count += 1;
            }
        }
        return count;
    }
}

File: python/depth-first-search/CountConnectedComponents.py
class Solution:

    def build_graph(self, edges, n):
        graph = []

        for i in range(n):
            graph.append(list())

        for edge in edges:
            graph[edge[0]].append(edge[1])
            graph[edge[1]].append(edge[0])

        return graph

    def solve(self, node, graph, visited):
        visited[node] = True
        for nbr in graph[node]:
            if not visited[nbr]:
                self.solve(nbr, graph, visited)        

    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = self.build_graph(edges, n)
        visited = [False] * n
        count = 0
        for src in range(n):
            if visited[src] is not True:
                self.solve(src, graph, visited)
                count += 1
        return count